### DISCUSSION TITLE
Instance/static method compile error

### DISCUSSION THREAD

**Discussion by jcouv** [dotnet/csharplang#1190] [General]
@karczk commented on [Thu Sep 21 2017](https://github.com/dotnet/roslyn/issues/22249)

**Steps to Reproduce**:

```
    class Program
    {
        static void Main(string[] args)
        {
            var instance = new Test();
            string input = "abc";
            instance.IsValid(input); // <- HERE
        }
    }

    class Test
    {
        public static bool IsValid(string value)
        {
            return true;
        }

        public bool IsValid(object value)
        {
            return true;
        }
    }
```

**Expected Behavior**:
Should compile, there is only one instance method and string is an object.

**Actual Behavior**:
Does not compile: 

> Member 'ConsoleApplication.Test.IsValid(string)' cannot be accessed with an instance reference; qualify it with a type name instead

---

@dpoeschl commented on [Thu Sep 21 2017](https://github.com/dotnet/roslyn/issues/22249#issuecomment-331242033)

This sounds more like a feature request for the language, since the error message is pretty clear that this is not expected to work. Feel free to port to https://github.com/dotnet/csharplang/issues if you'd like to propose a change to the language.

---

@jcouv commented on [Thu Dec 14 2017](https://github.com/dotnet/roslyn/issues/22249#issuecomment-351912334)

Moving language design discussions to csharplang repo.

**Comment by yaakov-h** [comment]
Is this an edge case of overload resolution?

**Comment by mattwar** [comment]
Yes.  The language is doing overload resolution on the full set of methods (not considering static vs member), picking the best (which is the static member in this case) and then generating the error.  

I would suppose the feature request would be to change overload resolution to consider the context (static vs member) when selecting the candidate methods.
